api_v3/services/UserAppRoleService.php

```diff
@@ -0,0 +1,196 @@
+{
+	/**
+	 * @param $serviceId
+	 * @param $serviceName
+	 * @param $actionName
+	 * @return void
+	 * @throws KalturaAPIException
+	 */
+	public function initService($serviceId, $serviceName, $actionName)
+	{
+		parent::initService($serviceId, $serviceName, $actionName);
+		
+		// Add Partner Ids to criteria for the following peers during execution of that service
+		$this->applyPartnerFilterForClass('UserRole');
+	}
+	
+	/**
+	 * Assign an application role for a user
+	 *
+	 * @action add
+	 *
+	 * @param KalturaUserAppRole $userAppRole
+	 * @return KalturaUserAppRole
+	 *
+	 * @throws KalturaAPIException
+	 * @throws PropelException
+	 * @throws Exception
+	 */
+	public function addAction(KalturaUserAppRole $userAppRole)
+	{
+		// todo: consider adding $master_partner_id = -11
+		// todo: if it's EP impersonated session, fetch user_role from 'global_partner' (0)
+		
+		$dbUserAppRole = $userAppRole->toInsertableObject();
+		$dbUserAppRole->save();
+		
+		$userAppRole = new KalturaUserAppRole();
+		$userAppRole->fromObject($dbUserAppRole, $this->getResponseProfile());
+		
+		return $userAppRole;
+	}
+	
+	/**
+	 * Update an application role for a user
+	 *
+	 * @action update
+	 *
+	 * @param string $userId
+	 * @param string $appGuid
+	 * @param KalturaUserAppRole $userAppRole
+	 * @return KalturaUserAppRole
+	 *
+	 * @throws KalturaAPIException
+	 * @throws PropelException
+	 * @throws Exception
+	 */
+	public function updateAction($userId, $appGuid, KalturaUserAppRole $userAppRole)
+	{
+		$dbUserAppRole = $this->getByUserAndAppGuid($userId, $appGuid);
+		$dbUserAppRole = $userAppRole->toUpdatableObject($dbUserAppRole);
+		$dbUserAppRole->save();
+		
+		$userAppRole = new KalturaUserAppRole();
+		$userAppRole->fromObject($dbUserAppRole, $this->getResponseProfile());
+		
+		return $userAppRole;
+	}
+	
+	/**
+	 * Get an application role for a user and app guid
+	 *
+	 * @action get
+	 *
+	 * @param string $userId the user id
+	 * @param string $appGuid the app-registry id
+	 * @return KalturaUserAppRole
+	 *
+	 * @throws KalturaAPIException
+	 * @throws PropelException
+	 * @throws kCoreException
+	 */
+	public function getAction($userId, $appGuid)
+	{
+		$dbUserAppRole = $this->getByUserAndAppGuid($userId, $appGuid);
+		
+		$userAppRole = new KalturaUserAppRole();
+		$userAppRole->fromObject($dbUserAppRole, $this->getResponseProfile());
+		
+		return $userAppRole;
+	}
+	
+	/**
+	 * List an application roles by filter and pager
+	 *
+	 * @action list
+	 *
+	 * @param KalturaUserAppRoleFilter $filter
+	 * @param KalturaFilterPager $pager
+	 * @return KalturaUserAppRoleListResponse
+	 *
+	 * @throws KalturaAPIException
+	 * @throws PropelException
+	 */
+	public function listAction(KalturaUserAppRoleFilter $filter = null, KalturaFilterPager $pager = null)
+	{
+		if (!$filter || !($filter->userIdEqual || $filter->userIdIn || $filter->userRoleIdEqual || $filter->userRoleIdIn || $filter->appGuidEqual || $filter->appGuidIn))
```

Comment:
I dont follow this limitation 
I think there are a few options here:

1. A request was made with user KS and you can only get a list of items that are specific to your user in any other case a user KS requests this for another user an exception should be thrown.

2. The request is with admin Ks and that you can provide an empty filter to fetch all.

You can say we have a limitation that in this service you can only do list if you either specify appGuid or userId in any other case fail it but this should be agreed with product.

Reply:
1. If we get a user ks, I override any userEqual or userIn and set the user from the ks on the filter, here:
https://github.com/kaltura/server/blob/3c0061350fcba6ef155e36f57ceb2f2cc14b927c/api_v3/lib/types/filters/KalturaUserAppRoleFilter.php#L94

That way, listAction always return a valid response (either 0 or > 0)

2. We cannot do that because there is not PID on that bridge table, so I must get at least one object_id that I can verify belongs to the PID